### PR TITLE
Fix: Add resolve for typeVariables in MethodSpec (#117)

### PR DESCRIPTION
**Pull request by PinkMoustacheMan**
## Before this PR
As described in the linked issue, Generics which are bounded by another generic type, are not correctly overridden by the `MethodSpec.overriding()` Method.

## After this PR
==COMMIT_MSG==
Fix: Add resolve for typeVariables in MethodSpec (#117 )
==COMMIT_MSG==

## Possible downsides?
None obvious to me.
There may be edge cases not covered by unit tests that are affected by this change.

### CODE REVIEW

**Comment by palantirtech** [comment]
Thanks for your interest in [palantir/javapoet](https://github.com/palantir/javapoet), @PinkMoustacheMan! Before we can accept your pull request, you need to sign our contributor license agreement - just visit https://cla.palantir.com/ and follow the instructions. Once you sign, I'll automatically update this pull request.

File: javapoet/src/main/java/com/palantir/javapoet/MethodSpec.java [hunk 1]
```diff
@@ -307,6 +309,18 @@ public static Builder overriding(ExecutableElement method, DeclaredType enclosin
...
+        for (int i = 0, size = builder.typeVariables.size(); i < size; i++) {
+            String name =
+                    resolvedTypeVariables.get(i).asElement().getSimpleName().toString();
+            List<TypeName> bounds = new ArrayList<>();
+            if (resolvedTypeVariables.get(i).getUpperBound().getKind() != TypeKind.NULL) {
+                bounds.add(TypeName.get(resolvedTypeVariables.get(i).getUpperBound()));
+            }
+            if (resolvedTypeVariables.get(i).getLowerBound().getKind() != TypeKind.NULL) {
+                bounds.add(TypeName.get(resolvedTypeVariables.get(i).getLowerBound()));
+            }
```
Comment (carterkozak) [inline javapoet/src/main/java/com/palantir/javapoet/MethodSpec.java]: I don't believe this statement is correct, the lower bound is used for things like `Collection<? super CharSequence>`, where `T extends CharSequence` is an upper bound, which is all that is supported for method type variables (as opposed to type references).

File: javapoet/src/main/java/com/palantir/javapoet/MethodSpec.java [hunk 2]
```diff
@@ -307,6 +309,18 @@ public static Builder overriding(ExecutableElement method, DeclaredType enclosin
             builder.parameters.set(
                     i, parameter.toBuilder(type, parameter.name()).build());
         }
+        for (int i = 0, size = builder.typeVariables.size(); i < size; i++) {
+            String name =
+                    resolvedTypeVariables.get(i).asElement().getSimpleName().toString();
```
Comment (carterkozak) [inline javapoet/src/main/java/com/palantir/javapoet/MethodSpec.java]: Would you mind extracting out the `resolvedTypeVariables.get(i)` into `TypeVariable resolvedTypeVariable = resolvedTypeVariables.get(i)` to keep the implementation a bit more concise?

File: javapoet/src/test/java/com/palantir/javapoet/MethodSpecTest.java [hunk 3]
```diff
@@ -227,6 +235,15 @@ public int compareTo(com.palantir.javapoet.MethodSpecTest.ExtendsOthers arg0) {
...
                         """);
+        exec = findFirst(methods, "doStuff");
+        method = MethodSpec.overriding(exec, classType, types).build();
+        assertThat(method.toString())
+                .isEqualTo(
+                        """
+                        @java.lang.Override
+                        public <S extends java.lang.Double> void doStuff(S input) {
+                        }
+                        """);
```
Comment (carterkozak) [inline javapoet/src/test/java/com/palantir/javapoet/MethodSpecTest.java:246]: Great test, thanks!

**Review by carterkozak — COMMENTED** [review]
Generally looks good, thank you for the contribution.

I'm terribly sorry for the delayed review!

File: javapoet/src/main/java/com/palantir/javapoet/MethodSpec.java (hunk 1 above)
Comment (PinkMoustacheMan) [reply javapoet/src/main/java/com/palantir/javapoet/MethodSpec.java]: I have removed the lower bound

File: javapoet/src/main/java/com/palantir/javapoet/MethodSpec.java (hunk 2 above)
Comment (PinkMoustacheMan) [reply javapoet/src/main/java/com/palantir/javapoet/MethodSpec.java]: Done

File: javapoet/src/main/java/com/palantir/javapoet/MethodSpec.java [hunk 4]
```diff
@@ -307,6 +309,15 @@ public static Builder overriding(ExecutableElement method, DeclaredType enclosin
             builder.parameters.set(
                     i, parameter.toBuilder(type, parameter.name()).build());
         }
+        for (int i = 0, size = builder.typeVariables.size(); i < size; i++) {
+            TypeVariable resolvedTypeVariable = resolvedTypeVariables.get(i);
+            String name = resolvedTypeVariable.asElement().getSimpleName().toString();
+            List<TypeName> bounds = new ArrayList<>();
+            if (resolvedTypeVariables.get(i).getUpperBound().getKind() != TypeKind.NULL) {
+                bounds.add(TypeName.get(resolvedTypeVariables.get(i).getUpperBound()));
```
Comment (carterkozak) [inline javapoet/src/main/java/com/palantir/javapoet/MethodSpec.java]:
```suggestion
            if (resolvedTypeVariable.getUpperBound().getKind() != TypeKind.NULL) {
                bounds.add(TypeName.get(resolvedTypeVariable.getUpperBound()));
```

**Review by carterkozak — APPROVED** [review]
